Sources/PackageCollections/Search/InMemoryPackageCollectionsSearch.swift

```diff
@@ -0,0 +1,361 @@
+    private let queue = DispatchQueue(label: "org.swift.swiftpm.InMemoryPackageCollectionsSearch", attributes: .concurrent)
+    
+    init(configuration: Configuration = .init()) {
+        self.configuration = configuration
+        self.packageTrie = Trie<CollectionPackage>()
+        self.targetTrie = Trie<CollectionPackage>()
+    }
+    
+    func index(collection: Model.Collection,
+               callback: @escaping (Result<Void, Error>) -> Void) {
+        self.queue.async {
+            let group = DispatchGroup()
+            
+            collection.packages.forEach { package in
+                group.enter()
+                
+                self.queue.async {
+                    defer { group.leave() }
+                    
+                    let document = CollectionPackage(collection: collection.identifier, package: package.reference.identity)
+        
+                    // This breaks up the URL into tokens because it contains punctuations, so when the
+                    // search query is a repository URL it will need to be tokenized the same way.
+                    // `searchPackages` does this, but not sure this is a good thing? We don't provide
+                    // options to NOT tokenize query string.
+                    self.index(text: package.repository.url, foundIn: document, to: self.packageTrie)
+                    // Index package identity without any transformation for `findPackage`
+                    self.index(text: package.reference.identity.description, foundIn: document, to: self.packageTrie, analyze: false)
+
+                    if let summary = package.summary {
+                        self.index(text: summary, foundIn: document, to: self.packageTrie)
+                    }
+                    if let keywords = package.keywords {
+                        keywords.forEach {
+                            self.index(text: $0, foundIn: document, to: self.packageTrie)
+                        }
+                    }
+                    package.versions.forEach { version in
+                        self.index(text: version.packageName, foundIn: document, to: self.packageTrie)
+                        version.products.forEach {
+                            self.index(text: $0.name, foundIn: document, to: self.packageTrie)
+                        }
+                        version.targets.forEach {
+                            self.index(text: $0.name, foundIn: document, to: self.packageTrie)
+                            // Target search is not tokenized - it's either exact match or prefix match
+                            self.index(text: $0.name, foundIn: document, to: self.targetTrie, analyze: false)
+                        }
+                    }
+                }
+            }
+            
+            group.notify(queue: self.queue) {
+                self.cacheLock.withLock {
+                    self.cache[collection.identifier] = collection
```

Comment:
for my understanding, so this will basically update the in-memory copy with the latest doc index?

Reply:
Something to consider is if we should store this info inside the trie, for it would simplify de/serialization I think. I need to study `SQLitePackageCollectionsStorage` some more, but the direction I am leaning towards is embed this inside `SQLitePackageCollectionsStorage` rather than making them separate strategies.